src/org/rascalmpl/util/maven/Artifact.java

```diff
@@ -130,70 +133,159 @@ private static boolean isVersionRange(String version) {
+            artifact.calculateClassPath(forScope, alreadyIncluded, resolvedVersions, rangedDeps, excludes, result, parser);
+        }
+    }
+
+    /**
+     * We will use the first version we encounter (breadth-first) and cache it.
+     * Note that caching is done modulo transitive exclusions.
+     * For ranged versions we just use the first suitable version for the first range we encounter.
+     */
+    private void calculateClassPath(Scope forScope, Set<ResolveKey> alreadySolved, Map<WithoutVersion, String> resolvedVersions, Map<WithoutVersion, SortedSet<String>> rangedDeps,
+        Set<WithoutVersion> excludes, ArrayList<Artifact> result, MavenParser parser) {
+        var nextLevel = new ArrayList<ResolveState>(dependencies.size());
         for (var d : dependencies) {
             var coordinate = d.getCoordinate();
-            var withoutVersion = coordinate.versionLess();
+
+            boolean alreadyResolved;
+
+            var versionLess = coordinate.versionLess();
 
             var version = coordinate.getVersion();
             if (isVersionRange(version)) {
                 try {
-                    version = ourResolver.findLatestMatchingVersion(coordinate.getGroupId(), coordinate.getArtifactId(), version);
-                    coordinate = new ArtifactCoordinate(coordinate.getGroupId(), coordinate.getArtifactId(), version, coordinate.getClassifier());
+                    version = ourResolver.findLatestMatchingVersion(coordinate.getGroupId(), coordinate.getArtifactId(),
+                        version);
+                    coordinate = new ArtifactCoordinate(coordinate.getGroupId(), coordinate.getArtifactId(), version,
+                        coordinate.getClassifier());
                 }
                 catch (UnresolvableModelException e) {
-                    messages = messages.append(Messages.error("Version range error: " + e.getMessage(), d.getPomLocation()));
+                    messages.append(Messages.error("Version range error: " + e.getMessage(), d.getPomLocation()));
                     continue;
                 }
 
-                var versions = rangedDeps.computeIfAbsent(withoutVersion, k -> new TreeSet<>());
+                var versions = rangedDeps.computeIfAbsent(versionLess, k -> new TreeSet<>());
                 if (versions.add(version) && versions.size() == 2) { // Only add a warning once
                     String effectiveVersion = versions.first();
-                    messages = messages.append(Messages.warning("Multiple version ranges found for " + withoutVersion + " are used. " 
-                        + "It is better to lock the desired version in your own pom to a specifick version, for example <version>["+ effectiveVersion + "]</version>", d.getPomLocation()));
+                    messages = messages.append(Messages.warning("Multiple version ranges found for " + versionLess
+                        + " are used. "
```

Comment:
nit: also here, the reformatting of this string does not make it more readable (and reviewable).